proposed/2023/NuGetAudit-without-nuget.org-as-packageSource.md

```diff
@@ -0,0 +1,190 @@
+However, it's not clear what customer benefit would result in being able to use different audit sources in different projects in the same solution or repo.
+My best guess is that the most likely scenario is customers using different settings when working in the office compared to working from home.
+However, we have not received much feedback that this is desirable for package sources, so it's unlikely that it's important for audit sources.
+
+## Prior Art
+
+<!-- What prior art, both good and bad are related to this proposal? -->
+<!-- Do other features exist in other ecosystems and what experience have their community had? -->
+<!-- What lessons from other communities can we learn from? -->
+<!-- Are there any resources that are relevant to this proposal? -->
+Although [`npm audit`'s help docs](https://docs.npmjs.com/cli/v10/commands/npm-audit?v=true) don't mention it, it appears to have a `--registry={url}` argument, as can be seen on [the Azure DevOps docs on npm audit](https://learn.microsoft.com/azure/devops/artifacts/npm/npm-audit?view=azure-devops&tabs=yaml).
+
+Pypi's `pip` has [an audit command](https://pypi.org/project/pip-audit/), and `--index-url` and `--extra-index-url` appear to enable using custom audit sources.
+
+Rust's cargo audit command has [configuration for the advisory DB URL](https://docs.rs/cargo-audit/latest/cargo_audit/config/struct.DatabaseConfig.html)
+
+## Unresolved Questions
+
+<!-- What parts of the proposal do you expect to resolve before this gets accepted? -->
+<!-- What parts of the proposal need to be resolved before the proposal is stabilized? -->
+<!-- What related issues would you consider out of scope for this proposal but can be addressed in the future? -->
+
+1. default value?
+
+   Should we add nuget.org as a default value, so that customers who remove nuget.org as a package source are more easily able to use NuGetAudit?
+   Would using nuget.org as a default audit source cause undue concern from customers who don't want to use nuget.org as a package source, and the new tooling version starts making network requests to nuget.org?
+
+   If we want to make nuget.org a default/automatic audit source, how?
+   NuGet has tried before to use a tracking file to automatically add nuget's v3 URL as a package source to user-profile nuget.config files that don't have any package sources.
+   However, this caused a security issue when customers wanted to remove nuget.org as a default package source, and this code incorrectly re-added it once (the customer could successfully remove nuget.org a second time).
+   When we removed this auto-update to mitigate the security concern, it exposed a bug where multiple other apps created user-profile nuget.config files without any package sources, and many customers assumed it was a bug with NuGet.
+
+   Should we hard-code nuget.org's default URL in NuGet.Configuration, and require customers to use `<clear/>` if they don't want any audit source?
```

Comment:
This feels weird. As a new feature, don't force adoption, and we can consider later how we might funnel customers into it?

Reply:
On the other hand, learning from our team's experience of trying to migrate from NuGet's V2 URL to the V3 URL, and how difficult it is to get existing developers learn about and adopt changes to the default nuget.config, I was looking for other ideas.

Additionally, while I'm not an expert in any other package system, I believe that for npm, pypi, cargo, and probably others, you don't need a configuration file that states the default package repository, or audit service url. My understanding it's hardcoded into the package manager, and if you don't want to use the official package source, then you need to explicitly configure that.

NuGet on the other hand is almost (but not quite) the opposite. nuget.org is only used if explicitly stated in config files, which NuGet will create when no user-profile config file exists, but this means that every developer has a config file, even when they're happy to use defaults, which means it's next to impossible for us to actually change defaults, since it requires changes to the config files.

And that's exactly the problem we had with the V2 to V3 URL migration. NuGet can't tell the difference between "customer who is using the default, so we can/should change" vs "customer explicitly wants to use this setting, which happens to be the old default".